[8.x] Backport 18513, 18694

ofrobots opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
8.x backport of #18513, #18694. @nodejs/lts this probably requires extra scrutiny as it backports a semver-minor deprecation – because async_hooks is experimental – to an LTS release. 

I don't have too strong feelings that the deprecation *must* be backported to 8.x, but the rest of the commit (addition of the `runInAsyncScope` API) would be good to backport so that users have a consistent async_hooks API available across node versions. I'd be open to dropping the deprecation, if people feel strongly.

There are very few, but non-zero, users of the async hooks embedder API at this point.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

EDIT: CI: https://ci.nodejs.org/job/node-test-pull-request/13814/

MylesBorins commented:
@ofrobots if the api is experimental aren't any changes considered semver-patch?

ofrobots commented:
@MylesBorins that's a good point. I'm fine with whatever @nodejs/lts decides on this.

ofrobots commented:
ping @nodejs/lts, how do we proceed with this one. 

I have rebased this PR. New CI: https://ci.nodejs.org/job/node-test-pull-request/14410/.

MylesBorins commented:
@ofrobots we can plan a semver-minor release for the next cycle of 8.x. For now I think this has to wait

ofrobots commented:
Done. New CI: https://ci.nodejs.org/job/node-test-pull-request/15042/

MylesBorins commented:
One more CI: https://ci.nodejs.org/job/node-test-pull-request/

refack commented:
Can one of the admins verify this patch?

ofrobots commented:
Ping @nodejs/async_hooks @nodejs/diagnostics  in case people have opinions on this backport given discussion about async-hooks being "quasi" experimental at the Diag WG meeting yesterday.

@nodejs/lts this has been open for a while.

MylesBorins commented:
One more CI, if green I'll land

https://ci.nodejs.org/job/node-test-pull-request/15464/

Next 8.x is going to be semver minor

ofrobots commented:
@mcollina can you clarify? You want to remove the deprecation from `master`? Those have long shipped as part of Node 10.

mcollina commented:
I mean remove the deprecated methods from master.

MylesBorins commented:
@mcollina are you asking for code changes to remove the deprecation in this pr?

mcollina commented:
I’d prefer to, but I won’t block this if the prevailing opinion is to deprecate in 8.

MylesBorins commented:
@ofrobots this needs a rebase, and could you respond to @mcollina's comment

ofrobots commented:
Rebased.

Removed runtime deprecation warnings, so this is a docs-only deprecation for the purposes of the back-port to 8.x.

New CI: https://ci.nodejs.org/job/node-test-pull-request/15540/

ofrobots commented:
Failures in the CI look unrelated. The only source change left in this PR is an addition of a method to the `AsyncResource` class which should not cause those `http2` and `timer` tests to fail. Flakes?

ofrobots commented:
Landed on `v8.x-staging` as 646445bfda...6696271837.
